File: templates/python/parser.py
class Parser(Grammar):
    def __init__(self, input, actions, types):
        self._input = input
        self._input_size = len(input)
        self._actions = actions
        self._types = types
        self._offset = 0
        self._cache = defaultdict(dict)
        self._failure = 0
        self._expected = []

    def parse(self):
        tree = self._read_{{root}}()
        if tree is not FAILURE and self._offset == self._input_size:
            return tree
        if not self._expected:
            self._failure = self._offset
            self._expected.append(({{{grammar}}}, '<EOF>'))
        raise ParseError(format_error(self._input, self._failure, self._expected))


class ParseError(SyntaxError):
    pass


def parse(input, actions=None, types=None):
    parser = Parser(input, actions, types)
    return parser.parse()

def format_error(input, offset, expected):
    lines = input.split('\n')
    line_no, position = 0, 0

    while position <= offset:
        position += len(lines[line_no]) + 1
        line_no += 1

    line = lines[line_no - 1]
    message = 'Line ' + str(line_no) + ': expected one of:\n\n'

    for pair in expected:
        message += '    - ' + pair[1] + ' from ' + pair[0] + '\n'

    number = str(line_no)
    while len(number) < 6:
        number = ' ' + number

    message += '\n' + number + ' | ' + line + '\n'
    message += ' ' * (len(line) + 10 + offset - position)
    return message + '^'

File: templates/cs/Parser.cs
using System.Collections.Generic;
using System.Collections;
using System;

namespace {{namespace}} {
    public class {{name}} : Grammar {
        public {{name}}(String input, Actions actions) {
            this.input = input;
            this.inputSize = input.Length;
            this.actions = actions;
            this.offset = 0;
            this.cache = new Dictionary<Label, Dictionary<int, CacheRecord>>();
            this.failure = 0;
            this.expected = new List<String[]>();
        }

        public static TreeNode parse(String input, Actions actions) {
            {{name}} parser = new {{name}}(input, actions);
            return parser.parse();
        }

        public static TreeNode parse(String input){
            return parse(input, null);
        }

        private static String formatError(String input, int offset, List<String[]> expected) {
            String[] lines = input.Split('\n');
            int lineNo = 0, position = 0;

            while (position <= offset) {
                position += lines[lineNo].Length + 1;
                lineNo += 1;
            }

            String line = lines[lineNo - 1];
            String message = "Line " + lineNo + ": expected one of:\n\n";

            foreach (String[] pair in expected) {
                message += "    - " + pair[1] + " from " + pair[0] + "\n";
            }

            String number = "" + lineNo;
            while (number.Length < 6) number = " " + number;
            message += "\n" + number + " | " + line + "\n";

            position -= line.Length + 10;

            while (position < offset) {
                message += " ";
                position += 1;
            }
            return message + "^";
        }

        private TreeNode parse(){
            TreeNode tree = _read_{{root}}();
            if (tree != FAILURE && offset == inputSize) {
                return tree;
            }
            if (expected.Count <= 0) {
                failure = offset;
                expected.Add(new String[] { {{{grammar}}}, "<EOF>" });
            }
            throw new ParseError(formatError(input, failure, expected));
        }
    }
}
